File: go-code/organizing_data/replace_derived_variable_with_query/after/customer-repository.go
package after

// STEP 1: Create a central place to store the customers
type CustomerRepository struct {
	customers map[string]*Customer
}

func NewCustomerRepository() *CustomerRepository {
	return &CustomerRepository{
		customers: make(map[string]*Customer),
	}
}

func (cr *CustomerRepository) RegisterCustomer(id string) Customer {
	if cr.customers[id] == nil {
		cr.customers[id] = NewCustomer(id)
	}

	return cr.FindCustomer(id)
}

func (cr *CustomerRepository) FindCustomer(id string) Customer {
	return *cr.customers[id]
}

File: java-code/src/main/java/refactoring/organizing_data/change_value_to_reference/after/CustomerRepository.java
package refactoring.organizing_data.change_value_to_reference.after;

import java.util.HashMap;
import java.util.Map;

//STEP1: Create a repository for customer so that we dont have to create customer with the same id
// all the time
public class CustomerRepository {
  private Map<String, Customer> customers = new HashMap<>();
  public Customer registerCustomer(String id) {
    if(customers.get(id) == null) {
      customers.put(id, new Customer(id));
    }

    return findCustomer(id);
  }

  public Customer findCustomer(String id) {
    return customers.get(id);
  }
}
